driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -90,28 +94,36 @@ class DefaultServerMonitor implements ServerMonitor {
             final InternalConnectionFactory internalConnectionFactory,
                          final ClusterConnectionMode clusterConnectionMode,
                          @Nullable final ServerApi serverApi,
+                         final boolean faas,
                          final Provider<SdamServerDescriptionManager> sdamProvider) {
         this.serverSettings = notNull("serverSettings", serverSettings);
         this.serverId = notNull("serverId", serverId);
         this.serverMonitorListener = singleServerMonitorListener(serverSettings);
         this.internalConnectionFactory = notNull("internalConnectionFactory", internalConnectionFactory);
         this.clusterConnectionMode = notNull("clusterConnectionMode", clusterConnectionMode);
         this.serverApi = serverApi;
+        this.faas = faas;
         this.sdamProvider = sdamProvider;
-        monitor = new ServerMonitorRunnable();
-        monitorThread = new Thread(monitor, "cluster-" + this.serverId.getClusterId() + "-" + this.serverId.getAddress());
-        monitorThread.setDaemon(true);
-        roundTripTimeMonitor = new RoundTripTimeRunnable();
-        roundTripTimeMonitorThread = new Thread(roundTripTimeMonitor,
-                "cluster-rtt-" + this.serverId.getClusterId() + "-" + this.serverId.getAddress());
-        roundTripTimeMonitorThread.setDaemon(true);
+        monitor = new ServerMonitor();
+        roundTripTimeMonitor = null;
         isClosed = false;
     }
 
     @Override
     public void start() {
-        monitorThread.start();
-        roundTripTimeMonitorThread.start();
+        monitor.start();
+    }
+
+    private void ensureRoundTripTimeMonitorStarted() {
+        lock.lock();
```

Comment:
Could you help me understand the reason for using a lock here?

Reply:
1. `DefaultServerMonitor.ensureRoundTripTimeMonitorStarted` is executed by the `ServerMonitor` thread, and may write to `roundTripTimeMonitor`.
2. `DefaultServerMonitor.close` is executed by an application thread, let's call it `A`.
3. `DefaultServerMonitor.close` implicitly checks if `roundTripTimeMonitor` is not `null`, and if it's not, calls `roundTripTimeMonitor.close`. This check-then-act logic is correct only if neither of the following can happen concurrently with this logic being executed:
    - `roundTripTimeMonitor` is being written to
    - `roundTripTimeMonitor` is being started
4. To guarantee that the aforementioned assumption holds, threads `ServerMonitor` and `A` must coordinate what they do<sup>1</sup>. For example, having them acquire the same exclusive lock guarantees that `ServerMonitor` cannot write to or start `roundTripTimeMonitor` concurrently with `A` executing its check-then-act logic.

Granted, the `roundTripTimeMonitor == null` check could have been moved out of the critical section in `ensureRoundTripTimeMonitorStarted` without introducing any correctness issues. However, doing so still has negative effects without yielding anything useful (other than flexing/posing?):

- It requires writing a lengthy comment explaining why `roundTripTimeMonitor == null` does not have to be executed as part of the critical section.
- It makes the code more fragile: as soon as we have another place that may write to `roundTripTimeMonitor` concurrently with `DefaultServerMonitor.ensureRoundTripTimeMonitorStarted` being executed (I don't see why we would do that, but that's besides the point), it's no longer correct to execute `roundTripTimeMonitor == null` outside of the critical section.

---

<sup>1</sup> Note how I only talk about the coordination here, and not the memory synchronization that affects observability of writes. This is because the latter could have been achieved without locking, and, therefore, is not relevant to the explanation.

Reply:
It's so good you asked about this code, because I have realized that `ensureRoundTripTimeMonitorStarted` must also check `isClosed`, as otherwise we still have a race condition between `ServerMonitor` starting `roundTripTimeMonitor` and `A` shutting it down. I'll fix this. This also does not invalidate anything I said in the previous comment.

Reply:
I also believe that the coordination described above can be done in a non-blocking way, and I remember doing so somewhere in our codebase previously. But I neither remember where it was, nor thing that would be the optimal approach here (I don't know whether the approach was optimal there).

Reply:
Thanks, @stIncMale, for the detailed explanation. Great that the race condition is caught.

[Optional] We could minimize acquiring the lock when the monitor has already started or closed by checking isClosed outside the synchronized block:

private void ensureRoundTripTimeMonitorStarted() {
    if (!isClosed) {
        withLock(lock, () -> {
            if (!isClosed && roundTripTimeMonitor == null) {
                roundTripTimeMonitor = new RoundTripTimeMonitor();
                roundTripTimeMonitor.start();
            }
        });
    }
}

Reply:
> We could minimize acquiring the lock when the monitor has already started ...

That requires moving `roundTripTimeMonitor == null` our of the block, which has the negative consequences described above, which I want to avoid. And I don't want to do the double check for `roundTripTimeMonitor`, because despite eliminating those negative consequences, int introduces another one: the second check from within the critical section would not be there for correctness (it's not needed), but to eliminate the aforementioned negative consequences. This means, the code will again need a lengthy comment describing why an totally unnecessary check is done the second time inside the critical section.

If we could gain a perceivable performance improvement from such a change, I would have probably done it despite those negative consequences, but it's clear even without measuring :trollface: that there will be no gains out of it.

> ... or closed by checking isClosed outside the synchronized block

Unless I am missing something, double checking `isClosed` saves us at best a single invocation of `lock.lock()` per instance of `DefaultServerMonitor` 😃

Reply:
I feel like doing the dumbest and simplest implementation here (in this particular case it does not seem to have any real negative consequences), instead of being a smart ass (there were situations when I wanted the latter).